cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,326 @@
+		if err != nil {
+			return err
+		}
+		fName := d.Name()
+		if d.IsDir() && fName != root { // skip any dir in the fs except root
+			return fs.SkipDir
+		}
+		info, err := d.Info()
+		if err != nil {
+			return err
+		}
+		if !info.Mode().IsRegular() || strings.Contains(fName, "..") {
+			return nil
+		}
+		// require one and only one file with name in the format
+		// `notation-{plugin-name}`
+		logger.Debugf("Processing file %s...", fName)
+		_, err = plugin.ExtractPluginNameFromFileName(fName)
+		if err != nil {
+			logger.Debugf("File name %s is not in format notation-{plugin-name}, skipped", fName)
+			return nil
+		}
+		rc, err := pluginFs.Open(path)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		if err := installPluginExecutable(ctx, fName, rc, force); err != nil {
+			return err
+		}
+		// on success, skip all remaining files
+		success = true
+		return fs.SkipAll
+	}); err != nil {
+		return err
+	}
+	if !success {
+		return ErrNoPluginExecutableFileWasFound
+	}
+	return nil
+}
+
+// installPluginFromTarGz extracts and untar a plugin tar.gz file, validates and
+// installs the plugin
+func installPluginFromTarGz(ctx context.Context, tarGzPath string, force bool) error {
+	logger := log.GetLogger(ctx)
+	rc, err := os.Open(tarGzPath)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	decompressedStream, err := gzip.NewReader(rc)
+	if err != nil {
+		return err
+	}
+	defer decompressedStream.Close()
+	tarReader := tar.NewReader(decompressedStream)
+	// require one and only one file with name in the format
+	// `notation-{plugin-name}`
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if !header.FileInfo().Mode().IsRegular() || strings.Contains(header.Name, "..") {
+			continue
+		}
+		fName := filepath.Base(header.Name)
+		logger.Debugf("Processing file %s...", fName)
+		_, err = plugin.ExtractPluginNameFromFileName(fName)
+		if err != nil {
+			logger.Infof("File name %s is not in format notation-{plugin-name}, skipped", fName)
+			continue
+		}
+		return installPluginExecutable(ctx, fName, tarReader, force)
+	}
+	return fmt.Errorf("no valid plugin file was found in %s. Plugin file name must in format notation-{plugin-name}", tarGzPath)
+}
+
+// installPluginExecutable extracts, validates, and installs a plugin executable
+// file
+func installPluginExecutable(ctx context.Context, fileName string, fileReader io.Reader, force bool) error {
+	tmpDir, err := os.MkdirTemp("", notationPluginTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to create notationPluginTmpDir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpFilePath := filepath.Join(tmpDir, fileName)
+	pluginFile, err := os.OpenFile(tmpFilePath, os.O_WRONLY|os.O_CREATE, 0700)
+	if err != nil {
+		return err
+	}
+	lr := &io.LimitedReader{
+		R: fileReader,
+		N: notationplugin.MaxPluginSourceBytes,
+	}
```

Comment:
Are we saying the size of plugin executable cannot be more than 256 MB (compressed or uncompressed doesnt matters)?